hal/src/common/dmac/dma_controller.rs

```diff
@@ -0,0 +1,293 @@
+                pub struct Channels(
+                    #(
+                        pub Channel<[<Ch N>], Uninitialized>,
+                    )*
+                );
+            }
+        });
+    };
+}
+
+with_num_channels!(define_channels_struct);
+
+/// Initialized DMA Controller
+pub struct DmaController {
+    dmac: DMAC,
+}
+
+#[bitfield]
+#[repr(u16)]
+pub struct PriorityLevelMask {
+    #[skip]
+    _reserved: B8,
+    /// Level 0
+    #[allow(dead_code)]
+    level0: bool,
+    /// Level 1
+    #[allow(dead_code)]
+    level1: bool,
+    /// Level 2
+    #[allow(dead_code)]
+    level2: bool,
+    /// Level 3
+    #[allow(dead_code)]
+    level3: bool,
+    #[skip]
+    _reserved: B4,
+}
+
+#[bitfield]
+#[repr(u32)]
+pub struct RoundRobinMask {
+    #[skip]
+    _reserved: B7,
+    /// Level 0
+    #[allow(dead_code)]
+    level0: bool,
+    #[skip]
+    _reserved: B7,
+    /// Level 1
+    #[allow(dead_code)]
+    level1: bool,
+    #[skip]
+    _reserved: B7,
+    /// Level 2
+    #[allow(dead_code)]
+    level2: bool,
+    #[skip]
+    _reserved: B7,
+    /// Level 3
+    #[allow(dead_code)]
+    level3: bool,
+}
+
+impl DmaController {
+    /// Return an immutable reference to the underlying DMAC object exposed by
+    /// the PAC.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because `DmaController` may expect certain
+    /// registers to retain a configuration. Messing with that configuration may
+    /// be unsafe.
+    pub unsafe fn dmac(&self) -> &DMAC {
+        &self.dmac
+    }
+
+    /// Return a mutable reference to the underlying DMAC object exposed by the
+    /// PAC.
+    ///
+    /// # Safety
+    ///
+    /// This function is unsafe because `DmaController` may expect certain
+    /// registers to retain a configuration. Messing with that configuration may
+    /// be unsafe.
+    pub unsafe fn dmac_mut(&mut self) -> &mut DMAC {
+        &mut self.dmac
+    }
+
+    /// Initialize the DMAC and return a DmaController object useable by
+    /// [`Transfer`](super::transfer::Transfer)'s. By default, all
+    /// priority levels are enabled unless subsequently disabled using the
+    /// `level_x_enabled` methods.
+    pub fn init(mut dmac: DMAC, _pm: &mut PM) -> Self {
+        // ----- Initialize clocking ----- //
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        {
+            // Enable clocking
+            _pm.ahbmask.modify(|_, w| w.dmac_().set_bit());
+            _pm.apbbmask.modify(|_, w| w.dmac_().set_bit());
+        }
+
+        Self::swreset(&mut dmac);
+
+        // SAFETY this is safe because we write a whole u32 to 32-bit registers,
+        // and the descriptor array addesses will never change since they are static.
+        // We just need to ensure the writeback and descriptor_section addresses
+        // are valid.
+        unsafe {
+            dmac.baseaddr
+                .write(|w| w.baseaddr().bits(DESCRIPTOR_SECTION.as_ptr() as u32));
```

Comment:
Does rustfmt do this? Wouldn't it be better to break at the beginning of the closure body? Maybe adding brackets would help?

Reply:
Yeah, that's rustfmt's fault. Even when adding a break with brackets, rustfmt will revert the formatting to above. I'm leaving it as is, otherwise rustfmt CI won't pass.